chore: Add first logging documentation (#10911)
hashgraph/hedera-services: platform-sdk/docs/base/logging/logging.md

## Hedera Logging API

### Introduction
Our custom logging facade is crafted to meet our unique requirements, ensuring compatibility with various software integrations. It acts as a bridge, redirecting events from our facade to SLF4J, providing flexibility for projects dependent on our modules.

### Architecture and Configuration
Our architecture does not dictate the format of logging output, allowing for a wide range of output options. This flexibility permits various configurations, from console outputs to integrations with systems like Graylog or Kibana.

### Features of Our Logging Facade
Our facade goes beyond mere message logging, encapsulating rich information within each logging event. Key features include:

1. **Support for Message Placeholders**
2. **Event Origin Tracking**
3. **Timestamp Inclusion**
4. **Thread Identification**
5. **Throwable Cause Handling**
6. **Support for Various Log Levels**
7. **Marker Support**
8. **Key-Value Based Metadata**
9. **Mapped Diagnostic Context (MDC) Support**

---

> ## Implementing the New Logging API

Reviewer: ```suggestion
## Using the New Logging API
```

### Setting Up the Logger

To integrate the new logging API within your Java application, begin by establishing a logger instance in your class. Here’s how you can set it up:

1. **Import the Logger Class**:
   Start by importing the `Loggers` class from the `com.swirlds.logging.api` package.

   ```java
   import com.swirlds.logging.api.Loggers;
   ```

2. **Create a Logger Instance**:
   In your class, declare a `private static final` logger instance. Use the `Loggers.getLogger()` method, passing your class (`MyClass.class`) as a parameter.

   ```java
   public class MyClass {

>        private static final Logger LOGGER = Loggers.getLogger(MyClass.class);

Author: We will provide a shorthand for IntelliJ or similar but this would introduce much overhead we want to avoid

### Logging Messages

Once you have set up the logger, you can proceed to log messages. The new API provides a straightforward and flexible way to log information. Here’s how to use it:

1. **Basic Logging**:
   For a simple log message, use the `info` method with a string argument.

   ```java
   LOGGER.info("Hello, world!");
   ```

2. **Logging with Placeholders**:
   The API supports placeholder syntax for dynamic message composition. Use curly braces `{}` as placeholders and pass the dynamic values as additional arguments.

   ```java
   LOGGER.info("Hello, {}!", "world");
   ```

   In this example, the `{}` placeholder in the string will be replaced by `"world"`, resulting in the log message `"Hello, world!"`.

---

## Custom Logging Configuration

The configuration syntax is designed for clarity and ease of use. It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.
For practicality, the `logging.properties` file should be set to reload every second, ensuring up-to-date logging configurations at all times.

> Regarding the format of the logging configuration, it should be user-friendly and self-explanatory. Here's an improved representation of the syntax for setting logging levels:

Reviewer: > Regarding the format of the logging configuration, it should be user-friendly and self-explanatory.

Too wordy... they are design goals...

```suggestion
 Here's a representation of the syntax for setting logging levels:
```

logging.level.com.swirlds.common.crypto.Signature = WARN
logging.level.com.hashgraph = WARN
```

This configuration demonstrates how to set a global default logging level (`INFO`). It also illustrates how to specify logging levels for packages and classes. For example, everything under `com.swirlds.common.crypto` is set to `DEBUG`, except for `com.swirlds.common.crypto.Signature`, which is explicitly set to `WARN`. Similarly, all loggers within `com.hashgraph` default to `WARN`. This approach ensures that loggers inherit the most specific level defined in the configuration, providing both flexibility and precision in logging management.

To provide developers with more control and flexibility in logging, we propose introducing filters for markers. This feature would allow developers to focus on log messages associated with specific markers, regardless of the logger's log level. If there's demand for such functionality among developers, the configuration for marker filters could be structured as follows:

```properties
# Marker filter configuration
logging.marker.CONFIG = ENABLED
logging.marker.CRYPTO = DISABLED
logging.marker.OTHER = DEFAULT

> ```

Reviewer: I'm not convinced about the style in which it is written...
I think that if the feature is available, it should explain what filters are, how to configure, what they provide after configuring, and leave out phrases like "we propose introducing...", " This feature would allow ..."; "If there's demand for such functionality among developers ..."
And if is not available, I wouldn't make this document the space for discussion.

A particularly useful feature of handlers is the `inheritLevels` property. This boolean property determines whether the handler should inherit the default level configurations set globally. By default, `inheritLevels` is set to `true`. However, it can be turned off to create a handler that focuses exclusively on a specific aspect of logging, such as a particular marker. For example, if you need a handler that only logs entries marked with the `CRYPTO` marker, your configuration would look like this:

```properties
# Handler specific for CRYPTO marker
logging.handler.CRYPTO_FILE.level = OFF
logging.handler.CRYPTO_FILE.marker.CRYPTO = ENABLED
```

> In this configuration, the `CRYPTO_FILE` handler is set to ignore the global log level settings (`level = OFF`) but is specifically enabled to log messages tagged with the `CRYPTO` marker (`marker.CRYPTO = ENABLED`). This setup allows for the creation of dedicated log files or outputs for specific types of log messages, providing a focused view that can be particularly useful in complex systems or during specific types of analysis.

Reviewer 2: `the creation of dedicated log files or outputs for specific types of log messages, providing a focused view`

Totally unrelated to this PR, and I'm not sure how exactly this is implemented today, but it seems that this is happening directly in the logger implementation somewhere. So I'm guessing that the code calling the logger might be penalized in terms of the performance to support all these features. It would be nice if these additional mechanisms (unconditional logging based on markers, redirecting some logs to additional destinations, etc.) would be happening in parallel (like on a different thread) or even completely offline to avoid performance degradation associated with logging which might be pretty high because in the end of the day the logging involves I/O.

If that already is the case, then it would be nice to mention this in the document to underscore that all these features don't incur any performance penalties.

## Hedera Logging API

### Introduction

> The Hedera Logging API is a custom logging facade which is crafted to meet our unique requirements, ensuring compatibility with various software integrations. It acts as a bridge, redirecting events from our facade to SLF4J, providing flexibility for projects dependent on our modules.

Reviewer 3: >  redirecting events from our facade to SLF4J
That is wrong. We do not want to forward anything to SLF4J. SLF4J should forward to our logging system. See diagram at https://github.com/hashgraph/hedera-services/issues/5424

### Architecture and Configuration
The architecture does not dictate the format of logging output, allowing for a wide range of output options. This flexibility permits various configurations, from console outputs to integrations with systems like Graylog or Kibana.

### Features of the Logging Facade
The facade goes beyond mere message logging, encapsulating rich information within each logging event. Key features include:

1. **Support for Message Placeholders**
2. **Event Origin Tracking**
3. **Timestamp Inclusion**
4. **Thread Identification**
5. **Throwable Cause Handling**
6. **Support for Various Log Levels**
7. **Marker Support**
8. **Key-Value Based Metadata**
9. **Mapped Diagnostic Context (MDC) Support**

---

## Using the Hedera Logging API

> 

Reviewer 3: Add Warning: DO NOT USE THIS API TODAY! WE STILL WORKING ON THE PERFORMANCE OF THE HANDLERS / OUTPUT.

### Setting Up the Logger

To integrate the new logging API within your Java application, begin by establishing a logger instance in your class. Here’s how you can set it up:

1. **Import the Logger Class**:
   Start by importing the `Loggers` class from the `com.swirlds.logging.api` package.

   ```java
   import com.swirlds.logging.api.Loggers;
   ```

2. **Create a Logger Instance**:
   In your class, declare a `private static final` logger instance. Use the `Loggers.getLogger()` method, passing your class (`MyClass.class`) as a parameter.

   ```java
   public class MyClass {

>        private static final Logger LOGGER = Loggers.getLogger(MyClass.class);

Reviewer 3: As far as I know we defined to use `logger` instead of `LOGGER`. Pls double check our coding guidelines.

   The API supports placeholder syntax for dynamic message composition. Use curly braces `{}` as placeholders and pass the dynamic values as additional arguments.

   ```java
   LOGGER.info("Hello, {}!", "world");
   ```

   In this example, the `{}` placeholder in the string will be replaced by `"world"`, resulting in the log message `"Hello, world!"`.

---

## Custom Logging Configuration

The configuration syntax is designed for clarity and ease of use. It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.
For practicality, the `logging.properties` file should be set to reload every second, ensuring up-to-date logging configurations at all times.

The format of the logging configuration is user-friendly and self-explanatory. Here's an example representation of the syntax for setting logging levels:

```properties
# Global default logging level
logging.level = INFO

# Specific logging levels for packages or classes
logging.level.com.swirlds.common.crypto = DEBUG
logging.level.com.swirlds.common.crypto.Signature = WARN
logging.level.com.hashgraph = WARN
```

> ### Levels

Reviewer 3: This headline should be above of the code sample

This configuration demonstrates how to set a global default logging level (`INFO`). It also illustrates how to specify logging levels for packages and classes. For example, everything under `com.swirlds.common.crypto` is set to `DEBUG`, except for `com.swirlds.common.crypto.Signature`, which is explicitly set to `WARN`. Similarly, all loggers within `com.hashgraph` default to `WARN`. This approach ensures that loggers inherit the most specific level defined in the configuration, providing both flexibility and precision in logging management.

To provide developers with more control and flexibility in logging, the configuration supports filters for markers. This feature allows developers to focus on log messages associated with specific markers, regardless of the logger's log level. Here's a sample configuration for marker filters:

```properties
# Marker filter configuration
logging.marker.CONFIG = ENABLED
logging.marker.CRYPTO = DISABLED
logging.marker.OTHER = DEFAULT
```

> ### Markers

Reviewer 3: This headline should be above line 93

In this configuration, the `logging.marker.NAME` pattern is used, where `NAME` represents the marker's name, and the value can be set to `ENABLED`, `DISABLED`, or `DEFAULT`. For example, `logging.marker.CONFIG = ENABLED` would ensure that all log messages tagged with the `CONFIG` marker are displayed, irrespective of their log level.

> A typical logging configuration file, incorporating these marker filters, might look like this:

Reviewer 3: Add a headline like "Using it all together" or so

```properties
# General logging level configuration
logging.level = INFO
logging.level.com.swirlds.common.crypto = DEBUG
logging.level.com.swirlds.common.crypto.Signature = WARN
logging.level.com.hashgraph = WARN

# Marker-specific logging configuration
logging.marker.CONFIG = ENABLED
```

### Handlers

To further refine our logging system, we have incorporated handlers for more granular control over logging behavior. Each handler can be distinctly named and configured using the prefix `logging.handler.NAME`, where `NAME` serves as a unique identifier. This structure allows for the application of handler-specific settings. For instance, `logging.handler.NAME.level` is used to set the logging level for a specific handler, ensuring that all previously discussed features such as marker filters and log level settings are compatible. This setup is instrumental in creating dedicated log files or outputs for specific types of log messages, offering a focused view that is particularly useful in complex systems or during targeted analyses. 

A key aspect of the Logging System design is its performance, emphasizing maximum efficiency with as low an impact as possible on the running system.

> A particularly useful feature of handlers is the `inheritLevels` property. This boolean property determines whether the handler should inherit the default level configurations set globally. By default, `inheritLevels` is set to `true`. However, it can be turned off to create a handler that focuses exclusively on a specific aspect of logging, such as a particular marker. For example, if you need a handler that only logs entries marked with the `CRYPTO` marker, your configuration would look like this:

Reviewer 3: Can you do one sentence per line? Otherwise the review will become very complexe since a comment can only be done for several sentences.

```properties
# Handler specific for CRYPTO marker
logging.handler.CRYPTO_FILE.level = OFF
logging.handler.CRYPTO_FILE.marker.CRYPTO = ENABLED
```

In this configuration, the `CRYPTO_FILE` handler is set to ignore the global log level settings (`level = OFF`) but is specifically enabled to log messages tagged with the `CRYPTO` marker (`marker.CRYPTO = ENABLED`). 

> ---

Reviewer 3: What is that line for?

## Test Support with `WithLoggingMirror`

> To improve testing, we've introduced the `WithLoggingMirror` annotation in JUnit 5. This annotation injects a `LoggingMirror` into test methods or classes, providing an isolated environment for logging event analysis. It ensures that tests with this annotation do not run in parallel, maintaining the integrity of each test's logging data.

Reviewer 3: Add link to our doc for junit extensions.

## Hedera Logging API

> [!WARNING]  
> The API is currently in develpment and is not optimized for production use.
 It is subject to change without notice.

### Introduction

> The Hedera Logging API is a custom logging facade which is crafted to meet our unique requirements, ensuring compatibility with various software integrations.

Reviewer 3: I would not say that our API is a logging facade. That would mean that we internally use a different logger. Better say Lib or API

> It acts as a bridge, redirecting events from eg SLF4J to our facade, providing flexibility for all dependencies used in the project.

Reviewer 3: First of all it acts as our logger that should be used in our code. Second it acts as a bridge. This now sounds like bridge is the main use case.

> It acts as a bridge, redirecting events from eg SLF4J to our facade, providing flexibility for all dependencies used in the project.

Reviewer 3: Add link to SLF4J

> 

Reviewer 3: I would add something like: "Our goal is to create a lightweight logging API that perfectly fits to our needs and performance in general"

> ### Architecture and Configuration

Reviewer 3: IO would not mix that 2 points. You have a full section about the config later. So concentrate on Architecture here.

> The architecture does not dictate the format of logging output, allowing for a wide range of output options.

Reviewer 3: Before the mentioned points I would add:
as any project of Base the logging is splitted in a public (api) and private (implementation) part and provides extension points next to a very basic functionality. For the logging lib Logging handlers and provider are extension points.
Another goal of the architecture is to not reinvent the wheel how logging in Java works. If you familiar with any Java logging lib you should be more or less ready to use our logger api.

This flexibility permits various configurations, from console outputs to integrations with systems like Graylog or Kibana.

> ### Features of the Logging Facade

Reviewer 3: Again: Not a facade

> ### Features of the Logging Facade

Reviewer 3: I would not add the headline but have the content under the architecture headline

The facade goes beyond mere message logging, encapsulating rich information within each logging event.
Key features include:

1. Support for Message Placeholders

> 2. Event Origin Tracking

Reviewer 3: limited to the class

3. Timestamp Inclusion
4. Thread Identification
5. Throwable Cause Handling
6. Support for Various Log Levels
7. Marker Support
8. Key-Value Based Metadata

> 9. Mapped Diagnostic Context (MDC) Support

Reviewer 3: I would say 8+9 are the same.

> 

Reviewer 3: Additional points:
- simple property based configuration
- Test Support

### Logging Messages

Once you have set up the logger, you can proceed to log messages.
The new API provides a straightforward and flexible way to log information.
Here’s how to use it:

1. **Basic Logging**:
   For a simple log message, use the `info` method with a string argument.

   ```java
   logger.info("Hello, world!");
   ```

2. **Logging with Placeholders**:
   The API supports placeholder syntax for dynamic message composition. 
   Use curly braces `{}` as placeholders and pass the dynamic values as additional arguments.

   ```java
   logger.info("Hello, {}!", "world");
   ```

   In this example, the `{}` placeholder in the string will be replaced by `"world"`, resulting in the log message `"Hello, world!"`.

## Custom Logging Configuration

The configuration syntax is designed for clarity and ease of use.
It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.

For practicality, the `logging.properties` file should be set to reload every second, ensuring up-to-date logging configurations at all times.

The format of the logging configuration is user-friendly and self-explanatory.

> Here's an example representation of the syntax for setting logging levels:

Reviewer 3: Move that line below the headline

       private static final Logger logger = Loggers.getLogger(MyClass.class);
   }
   ```

   This statement creates a logger specifically for `MyClass`, facilitating targeted logging within this class.

### Logging Messages

Once you have set up the logger, you can proceed to log messages.
The new API provides a straightforward and flexible way to log information.
Here’s how to use it:

1. **Basic Logging**:
   For a simple log message, use the `info` method with a string argument.

   ```java
   logger.info("Hello, world!");
   ```

2. **Logging with Placeholders**:
   The API supports placeholder syntax for dynamic message composition. 
   Use curly braces `{}` as placeholders and pass the dynamic values as additional arguments.

   ```java
   logger.info("Hello, {}!", "world");
   ```

   In this example, the `{}` placeholder in the string will be replaced by `"world"`, resulting in the log message `"Hello, world!"`.

## Custom Logging Configuration

The configuration syntax is designed for clarity and ease of use.
It allows setting a global default logging level and specific levels for packages or classes, with an option to introduce filters for markers.

> For practicality, the `logging.properties` file reload every second, ensuring up-to-date logging configurations at all times.

Reviewer 4: nit: `reload` => `will reload`

# General logging level configuration
logging.level = INFO
logging.level.com.swirlds.common.crypto = DEBUG
logging.level.com.swirlds.common.crypto.Signature = WARN
logging.level.com.hashgraph = WARN

# Marker-specific logging configuration
logging.marker.CONFIG = ENABLED
```

### Handlers

To further refine our logging system, we have incorporated handlers for more granular control over logging behavior. 
Each handler can be distinctly named and configured using the prefix `logging.handler.NAME`, where `NAME` serves as a unique identifier. 
This structure allows for the application of handler-specific settings. 
For instance, `logging.handler.NAME.level` is used to set the logging level for a specific handler, ensuring that all previously discussed features such as marker filters and log level settings are compatible. 
This setup is instrumental in creating dedicated log files or outputs for specific types of log messages, offering a focused view that is particularly useful in complex systems or during targeted analyses. 

A key aspect of the Logging System design is its performance, emphasizing maximum efficiency with as low an impact as possible on the running system.

> A particularly useful feature of handlers is the `inheritLevels` property.

Reviewer 4: The example doesn't show using `inheritLevels`.  Do we need a separate example for this?
